src/main/main.c

```diff
@@ -591,6 +605,52 @@ static int poll_setup(struct re *re)
 }
 
 
+static int fhs_update(struct re *re, struct fhs **fhsp, re_sock_t fd,
+		      int flags, fd_h *fh, void *arg)
+{
+	struct fhs *fhs = NULL;
+	struct le *le = hash_lookup(re->fhl, (uint32_t)fd, fhs_lookup, &fd);
```

Comment:
Since POSIX defines to reuse the lowest-numbered file descriptor, this should be fine. Also on windows high value fd numbers are very rare.

I think we have two practical use cases:
- Few (one or maybe 4) re event loop threads with many fds (>30000) (maxfds/bsize should be for best lookup performance equal to RLIMIT_NOFILE). memory overhead for hash_list is not a problem.
- Many (maybe thousands) event loop threads with some fds (<128), keep maxfds low to reduce memory usage.

In both cases the hash lookup with fd as hash key (`key & (h->bsize-1)`) index should be fine, right?

Reply:
In the "many event loops" case it is possible that fds have values with a period of `maxfds`, which would turn hash map into a list.

Also, not all fds are used with libre, meaning that even in the "many fds in an event loop" case you're not guaranteed to have even distribution of fd values. Though admittedly here a large skew is less likely.

I think a better hash function is still needed.

Reply:
I`m open for a good suggestion. But looks not to bad with quick worst case random u16 values:

```c
	int bsize = 128;
	for (int i = 0; i < 128; i++) {
		int fd = rand_u16();
		re_printf("%u\n", (fd & (bsize-1)));
	}
```

- 44% -> 1 list entry
- 34% -> 2 list entries
- 19% -> 3 list entries
- 3% -> 4 list entries

This can be improved with `bsize = 512` (~20% 2 list entries and 80% single list entry). A empty hash list with `bsize = 512` is only 8208 bytes and with 8192 threads this is only 64 MB overhead.

We can make `bsize` independent from maxfds.

Reply:
There are a few good candidates [here](https://en.wikipedia.org/wiki/List_of_hash_functions#Non-cryptographic_hash_functions). I would probably use something like simplified [Jenkins hash function](https://en.wikipedia.org/wiki/Jenkins_hash_function) (that is, just the mixing part, without the loop) or one round of [MurmurHash](https://en.wikipedia.org/wiki/MurmurHash), without mixing in the length. It doesn't need to be cryptographically strong, just something that reasonably well mixes the higher bits into the lower bits that are used by the hash table.

Reply:
Also, [here](http://www.burtleburtle.net/bob/hash/doobs.html)'s a good survey of hash functions.

Reply:
With our `hash_fast()` and `hash_joaat()` I see no real difference in distribution. But I will try to look closer.

Reply:
Which range of fd values do you expect? It should be safe to assume that these much lower than max uint32_t right? I see no benefit in hashing this value. Can you provide a example?

Reply:
Given my previous estimates, fds that differ in value by more than 128 or 512 are quite possible. Though the total number of fds per event loop won't be very large (I'm guessing, around ten or so on average).